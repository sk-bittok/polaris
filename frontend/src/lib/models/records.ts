export interface ProductionRecord {
	id: number;
	livestockName: string;
	animalPid: string;
	productType: string;
	unit: string;
	quantity: string;
	notes?: string;
	quality?: string;
	recordDate?: Date;
	createdAt: Date;
	createdBy: string;
	createdByName: string;
	updatedAt: Date;
}

export interface NewProductionRecord {
	id: number;
	animal_pid: string;
	organisation_pid: string;
	product_type: string;
	quantity: number;
	unit: string;
	record_date?: string;
	quality?: string;
	notes?: string;
	created_by: string;
	created_at: Date;
	updated_at: Date;
}

export interface HealthRecordResponse {
	id: number;
	organisationPid: string;
	organisationName: string;
	animalName: string;
	animalPid: string;
	animalTagId: string;
	recordType: string;
	recordDate: Date;
	description: string;
	treatment: string;
	medicine?: string;
	dosage?: string;
	cost?: number;
	notes?: string;
	performedBy?: string;
	createdBy: string;
	createdByName: string;
	createdAt: Date;
	updatedAt: Date;
}

export interface HealthRecord {
	id: number;
	organisation_pid: string;
	animal_pid: string;
	record_type: string;
	record_date: Date;
	description: string;
	treatment: string;
	medicine?: string;
	dosage?: string;
	cost?: number;
	notes?: string;
	performed_by?: string;
	created_by: string;
	created_at: Date;
	updated_at: Date;
}

export interface WeightRecord {
	id: number;
	organisation_pid: string;
	animal_pid: string;
	mass: number;
	record_date: string;
	notes?: string;
	created_by: string;
	created_at: Date;
	updated_at: Date;
}

export interface WeightRecordResponse {
	id: number;
	organisationPid: string;
	organisationName: string;
	animalPid: string;
	animalName: string;
	animalTagId: string;
	mass: number;
	recordDate: string;
	notes?: string;
	createdBy: string;
	createdByName: string;
	created_at: Date;
	updated_at: Date;
}
