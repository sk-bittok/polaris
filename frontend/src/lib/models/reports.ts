interface LivestockSummary {
	pid: string;
	organisationPid: string;
	id: number;
	total: number;
	males: number;
	females: number;
	unkownGender: number;
	active: number;
	transferred: number;
	sold: number;
	deceased: number;
	species: number;
	breeds: number;
	totalPurchasedValue: number;
	createdAt: Date;
	updatedAt: Date;
}

export type { LivestockSummary };
