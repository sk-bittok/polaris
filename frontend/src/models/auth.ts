import { z } from "zod";

export enum SubscriptionType {
    Basic = "basic",
    Business = "business",
    Enterprise = "enterprise",
}

export const companyFormSchema = z.object({
    companyName: z.string().min(2, { message: "Name requires 2 characters" })
        .max(200, { message: "Name is too long" }),
    companyAddress: z.string().max(200, { message: "Address is too long" })
        .nullable(),
    companyPhone: z.string().min(10, { message: "Phone is too short" })
        .max(12, { message: "Phone is too long" }).nullable(),
    companyEmail: z.string().email({ message: "Invalid email" }).nullable(),
    subscriptionType: z.enum(["basic", "business", "enterprise"], {
        message: "Basic, business and enterprise only",
    }),
});

export const userFormSchema = z.object({
    email: z.string().email({ message: "Invalid email" }),
    firstName: z.string().min(2, { message: "Name is too short" })
        .max(50, { message: "Name is too long" }),
    lastName: z.string().min(2, { message: "Name is too short" }).max(50, {
        message: "Name is too long",
    }),
    password: z.string().min(8, { message: "Password requires 8 characters" })
        .max(48, { message: "Password must be under 48 chars" }),
    confirmPassword: z.string(),
})
    .refine((data) => data.confirmPassword === data.password, {
        message: "Passwords must match",
        path: ["confirmPassword"],
    });

export const loginFormSchema = z.object({
    email: z.string().email({ message: "Invalid email" }),
    password: z.string().min(2, { message: "Password too short" })
        .max(100, { message: "Password too long" }),
});

export type CompanyForm = z.infer<typeof companyFormSchema>;
export type UserForm = z.infer<typeof userFormSchema>;

export type RegisterOrgAndUser = {
    organisation: CompanyForm;
    user: UserForm;
};

export type LoginForm = z.infer<typeof loginFormSchema>;
