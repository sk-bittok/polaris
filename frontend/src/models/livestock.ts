
export interface Breed {
  id: number;
  organisationPid?: string;
  specie: Category;
  name: string;
  maleWeightRange?: string;
  femaleWeightRange?: string;
  gestationPeriod?: string;
  description?: string;
  createdAt: string;
  isSystemDefined: boolean;
}

export enum Category {
  Cattle = "cattle",
  Sheep = "sheep",
  Goat = "goats",
  Chicken = "chicken",
  Pig = "pigs"
}

export enum Gender {
  Male = "male",
  Female = "female",
  Unkown = "unkown",
}

export enum Status {
  Active = "active",
  Sold = "sold",
  Deceased = "deceased",
  Transferred = "transferred"
}

export interface RegisterLivestock {
  tagId: string;
  name: string;
  gender: string;
  status: string;
  breed: string;
  specie: string;
  dateOfBirth?: string | Date | null;
  weightAtBirth?: number | null;
  maleParentId?: string | null;
  femaleParentId?: string | null;
  purchasePrice?: number | null;
  purchaseDate?: string | Date | null;
  currentWeight?: number | null;
  notes?: string | null;
};

export interface UpdateLivestock {
  tagId?: string | null;
  name?: string | null;
  gender?: string | null;
  status?: string | null;
  breed?: string | null;
  specie?: string | null;
  dateOfBirth?: string | Date | null;
  weightAtBirth?: number | null;
  maleParentId?: string | null;
  femaleParentId?: string | null;
  purchasePrice?: number | null;
  purchaseDate?: string | Date | null;
  currentWeight?: number | null;
  notes?: string | null;
};

export interface Livestock {
  id: number;
  pid: string;
  organisationName: string;
  tagId: string;
  name: string;
  breedName: string;
  specieName: Category;
  dateOfBirth?: Date;
  gender: Gender;
  parentMaleName?: string;
  parentFemaleName?: string;
  status: Status;
  parentMaleTagId?: string;
  parentFemaleTagId?: string;
  purchaseDate?:  Date;
  purchasePrice?: string;
  weightAtBirth?: string;
  currentWeight?: string;
  notes?: string;
  createdByName: string;
  createdAt: Date;
  updatedAt: Date;
}

