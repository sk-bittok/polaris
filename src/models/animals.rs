#![allow(clippy::missing_errors_doc)]
#![allow(clippy::too_many_lines)]

use async_trait::async_trait;
use chrono::{DateTime, FixedOffset, NaiveDate};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use sqlx::{Encode, Executor, Postgres, postgres::PgQueryResult, prelude::FromRow};
use uuid::Uuid;

use crate::seed::Seedable;

use super::{
    ModelError, ModelResult,
    dto::{RegisterAnimal, UpdateAnimal},
};

#[derive(Debug, Deserialize, Serialize, Encode, FromRow)]
#[serde(rename_all = "camelCase")]
pub struct AnimalCleaned {
    pub(crate) id: i32,
    pub(crate) pid: Uuid,
    pub(crate) organisation_pid: Uuid,
    pub(crate) organisation_name: String,
    pub(crate) tag_id: String,
    pub(crate) name: String,
    pub(crate) specie_name: String,
    pub(crate) breed_name: String,
    pub(crate) date_of_birth: Option<NaiveDate>,
    pub(crate) gender: String,
    pub(crate) parent_female_name: Option<String>,
    pub(crate) parent_female_tag_id: Option<String>,
    pub(crate) parent_female_id: Option<Uuid>,
    pub(crate) parent_male_name: Option<String>,
    pub(crate) parent_male_tag_id: Option<String>,
    pub(crate) parent_male_id: Option<Uuid>,
    pub(crate) status: String,
    pub(crate) purchase_date: Option<NaiveDate>,
    pub(crate) purchase_price: Option<Decimal>,
    pub(crate) weight_at_birth: Option<Decimal>,
    pub(crate) current_weight: Option<Decimal>,
    pub(crate) notes: Option<String>,
    pub(crate) created_by: Uuid,
    pub(crate) created_by_name: String,
    pub(crate) created_at: DateTime<FixedOffset>,
    pub(crate) updated_at: DateTime<FixedOffset>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct AnimalQuery {
    pub specie: Option<String>,
    pub breed: Option<String>,
    pub purchase_date: Option<NaiveDate>,
}

#[derive(Debug, Deserialize, FromRow, Encode, Serialize)]
pub struct Animal {
    pub(crate) id: i32,
    pub(crate) pid: Uuid,
    pub(crate) organisation_pid: Uuid,
    pub(crate) tag_id: String,
    pub(crate) name: String,
    pub(crate) specie_id: i32,
    pub(crate) breed_id: i32,
    pub(crate) date_of_birth: Option<NaiveDate>,
    pub(crate) gender: String,
    pub(crate) parent_female_id: Option<Uuid>,
    pub(crate) parent_male_id: Option<Uuid>,
    pub(crate) status: String,
    pub(crate) purchase_date: Option<NaiveDate>,
    pub(crate) purchase_price: Option<Decimal>,
    pub(crate) weight_at_birth: Option<Decimal>,
    pub(crate) current_weight: Option<Decimal>,
    pub(crate) notes: Option<String>,
    pub(crate) created_by: Uuid,
    pub(crate) created_at: DateTime<FixedOffset>,
    pub(crate) updated_at: DateTime<FixedOffset>,
}

const SELECT_QUERY: &str = "
            SELECT 
                a.id,
                a.pid,
                a.organisation_pid,
                o.name AS organisation_name,
                a.tag_id,
                a.name,
                b.name  AS  breed_name,
                s.name  AS  specie_name,
                a.gender,
                a.date_of_birth,
                a.current_weight,
                a.weight_at_birth,
                a.parent_female_id,
                f.tag_id  AS  parent_female_tag_id,
                g.name AS parent_female_name,
                a.parent_male_id,
                m.tag_id  AS  parent_male_tag_id,
                n.name AS parent_male_name,
                a.status,
                a.created_at,
                a.updated_at,
                a.purchase_date,
                a.purchase_price,
                a.notes,
                a.created_by,
                CONCAT(u.first_name, ' ', u.last_name) AS created_by_name
            FROM
                animals a
            LEFT JOIN
                organisations o ON a.organisation_pid = o.pid
            LEFT JOIN
                breeds b ON a.breed_id = b.id
            LEFT JOIN
                species s ON a.specie_id = s.id
            LEFT JOIN
                animals f ON a.parent_female_id = f.pid
            LEFT JOIN
                animals g ON a.parent_female_id = g.pid
            LEFT JOIN
                animals m ON a.parent_male_id = m.pid
            LEFT JOIN
                animals n ON a.parent_male_id = n.pid
            LEFT JOIN
                users u ON a.created_by = u.pid
            WHERE a.organisation_pid = $1
            ";

fn select_query(conditions: &str) -> String {
    format!("{SELECT_QUERY} {conditions}")
}

impl Animal {
    pub async fn find_all<'e, C>(
        db: C,
        org_pid: Uuid,
        conditions: &AnimalQuery,
    ) -> ModelResult<Vec<AnimalCleaned>>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let mut query = sqlx::query_as::<_, AnimalCleaned>(Box::leak(
            select_query("ORDER BY a.created_at DESC").into_boxed_str(),
        ))
        .bind(org_pid);

        if let Some(specie) = &conditions.specie {
            query = sqlx::query_as::<_, AnimalCleaned>(Box::leak(
                select_query("AND s.name ILIKE $2\nORDER BY a.created_at DESC").into_boxed_str(),
            ))
            .bind(org_pid)
            .bind(format!("%{}%", specie.as_str()));
        }

        if let Some(breed) = &conditions.breed {
            query = sqlx::query_as::<_, AnimalCleaned>(Box::leak(
                select_query("AND b.name ILIKE $2\nORDER BY a.created_at DESC").into_boxed_str(),
            ))
            .bind(org_pid)
            .bind(format!("%{}%", breed.as_str()));
        }

        query.fetch_all(db).await.map_err(Into::into)
    }

    pub async fn find_by_id<'e, C>(db: C, org_pid: Uuid, id: Uuid) -> ModelResult<AnimalCleaned>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query_as::<_, AnimalCleaned>(Box::leak(
            select_query("AND a.pid = $2").into_boxed_str(),
        ))
        .bind(org_pid)
        .bind(id)
        .fetch_optional(db)
        .await?
        .ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn find_by_tag_id<'e, C>(
        db: &C,
        org_pid: Uuid,
        tag_id: &str,
    ) -> ModelResult<AnimalCleaned>
    where
        for<'a> &'a C: Executor<'e, Database = Postgres>,
    {
        let item = sqlx::query_as::<_, AnimalCleaned>(Box::leak(
            select_query("AND a.tag_id = $2").into_boxed_str(),
        ))
        .bind(org_pid)
        .bind(tag_id.trim())
        .fetch_optional(db)
        .await?;

        item.ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn register<'e, C>(
        db: C,
        org_pid: Uuid,
        user_pid: Uuid,
        params: &RegisterAnimal<'_>,
    ) -> ModelResult<Self>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let purchase_price = params.purchase_price.map(|price| Decimal::new(price, 2));
        let current_weight = params.current_weight.map(|mass| Decimal::new(mass, 2));
        let birth_weight = params.weight_at_birth.map(|mass| Decimal::new(mass, 2));
        let birth_date = params
            .date_of_birth
            .as_ref()
            .map(|date| {
                let dob = date.split('T').collect::<Vec<_>>();
                Ok::<_, ModelError>(NaiveDate::parse_from_str(dob[0], "%Y-%m-%d")?)
            })
            .transpose()?;
        let purchase_date = params
            .purchase_date
            .as_ref()
            .map(|date| {
                let pd = date.split('T').collect::<Vec<_>>();
                Ok::<_, ModelError>(NaiveDate::parse_from_str(pd[0], "%Y-%m-%d")?)
            })
            .transpose()?;

        let query = sqlx::query_as::<_, Self>(
            "
            INSERT INTO animals
            (
                organisation_pid,
                tag_id,
                breed_id,
                specie_id,
                name,
                gender,
                date_of_birth,
                status,
                parent_female_id,
                parent_male_id,
                purchase_date,
                purchase_price,
                weight_at_birth,
                current_weight,
                notes,
                created_by
            )
            VALUES 
            (
                $1,
                $2,
                (SELECT id FROM breeds b WHERE b.name ILIKE $3 AND (b.is_system_defined = TRUE OR b.organisation_pid = $1)),
                (SELECT id FROM species s WHERE s.name LIKE $4),
                $5,
                $6,
                $7,
                $8,
                CASE
                    WHEN $9 IS NOT NULL THEN (SELECT pid FROM animals WHERE tag_id = $9 AND organisation_pid = $1)
                    ELSE NULL
                END,
                CASE
                    WHEN $10 IS NOT NULL THEN (SELECT pid FROM animals WHERE tag_id = $10 AND organisation_pid = $1)
                    ELSE NULL
                END,
                $11,
                $12,
                $13,
                $14,
                $15,
                $16
            )
            RETURNING *
            ",
        )
        .bind(org_pid)
        .bind(params.tag_id.as_ref())
        .bind(format!("%{}%", params.breed.as_ref()))
        .bind(format!("%{}%" ,params.specie.as_ref()))
        .bind(params.name.as_ref())
        .bind(params.gender.to_string())
        .bind(birth_date)
        .bind(params.status.as_ref())
        .bind(params.female_parent_id.as_ref())
        .bind(params.male_parent_id.as_ref())
        .bind(purchase_date)
        .bind(purchase_price)
        .bind(birth_weight)
        .bind(current_weight)
        .bind(params.notes.as_deref())
        .bind(user_pid)
        .fetch_one(db)
        .await?;

        Ok(query)
    }

    pub async fn update_by_id<'e, C>(
        db: &C,
        params: &UpdateAnimal<'_>,
        org_pid: Uuid,
        id: Uuid,
    ) -> ModelResult<Self>
    where
        for<'a> &'a C: Executor<'e, Database = Postgres>,
    {
        let item = Self::find_by_id(db, org_pid, id).await?;

        let tag_id = params
            .tag_id
            .as_ref()
            .map_or(item.tag_id.to_string(), ToString::to_string);
        let name = params
            .name
            .as_ref()
            .map_or(item.name.to_string(), ToString::to_string);
        let breed_name = params
            .breed
            .as_ref()
            .map_or(item.breed_name.to_string(), ToString::to_string);
        let specie_name = params
            .specie
            .as_ref()
            .map_or(item.specie_name.to_string(), ToString::to_string);
        let notes = params
            .notes
            .as_ref()
            .map_or(item.notes.clone(), |notes| Some(notes.to_string()));
        let gender = params
            .gender
            .as_ref()
            .map_or(item.gender.to_string(), ToString::to_string);
        let status = params
            .status
            .as_ref()
            .map_or(item.status.to_string(), ToString::to_string);
        let parent_female_tag_id = params
            .female_parent_id
            .as_ref()
            .map_or(item.parent_female_tag_id.clone(), |id| Some(id.to_string()));
        let parent_male_tag_id = params
            .male_parent_id
            .as_ref()
            .map_or(item.parent_male_tag_id.clone(), |id| Some(id.to_string()));
        let current_weight = params
            .current_weight
            .map_or(item.current_weight, |mass| Some(Decimal::new(mass, 2)));
        let weight_at_birth = params
            .weight_at_birth
            .map_or(item.weight_at_birth, |mass| Some(Decimal::new(mass, 2)));
        let purchase_price = params
            .purchase_price
            .map_or(item.purchase_price, |price| Some(Decimal::new(price, 2)));
        let date_of_birth =
            params
                .date_of_birth
                .as_ref()
                .map_or(Ok(item.date_of_birth), |date| {
                    let date = date.split('T').collect::<Vec<_>>();
                    Ok::<_, ModelError>(Some(NaiveDate::parse_from_str(date[0], "%Y-%m-%d")?))
                })?;
        let purchase_date =
            params
                .purchase_date
                .as_ref()
                .map_or(Ok(item.purchase_date), |date| {
                    let date = date.split('T').collect::<Vec<_>>();
                    Ok::<_, ModelError>(Some(NaiveDate::parse_from_str(date[0], "%Y-%m-%d")?))
                })?;

        let query = sqlx::query_as::<_, Self>(
            "
            UPDATE animals
            SET
                tag_id = $3,
                breed_id = (SELECT id FROM breeds b WHERE b.name ILIKE $4 AND (b.is_system_defined = true OR organisation_pid = $2 )),
                specie_id = (SELECT id FROM species s WHERE s.name ILIKE $5),
                name = $6,
                gender = $7,
                date_of_birth = $8,
                status = $9,
                parent_female_id =
                    CASE
                        WHEN $10 IS NOT NULL THEN (SELECT pid FROM animals WHERE tag_id = $10 AND organisation_pid = $2)
                        ELSE NULL
                    END,
                parent_male_id = 
                    CASE
                        WHEN $11 IS NOT NULL THEN (SELECT pid FROM animals WHERE tag_id = $11 AND organisation_pid = $2)
                        ELSE NULL
                    END,
                purchase_date = $12,
                purchase_price = $13,
                weight_at_birth = $14,
                current_weight = $15,
                notes = $16
            WHERE pid = $1 AND organisation_pid = $2
            RETURNING *
            ",
        )
        .bind(id)
        .bind(org_pid)
        .bind(tag_id)
        .bind(format!("%{breed_name}%"))
        .bind(format!("%{specie_name}%"))
        .bind(name)
        .bind(gender)
        .bind(date_of_birth)
        .bind(status)
        .bind(parent_female_tag_id)
        .bind(parent_male_tag_id)
        .bind(purchase_date)
        .bind(purchase_price)
        .bind(weight_at_birth)
        .bind(current_weight)
        .bind(notes)
        .fetch_one(db)
        .await?;

        Ok(query)
    }

    pub async fn seed(db: &sqlx::PgPool, path: &str) -> ModelResult<()> {
        let animals = Self::load_file(path).await?;

        Self::seed_data(db, &animals).await
    }

    pub async fn delete_by_id<'e, C>(db: C, org_id: Uuid, id: Uuid) -> ModelResult<PgQueryResult>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query("DELETE FROM animals WHERE pid = $1 AND organisation_pid = $2")
            .bind(id)
            .bind(org_id)
            .execute(db)
            .await
            .map_err(Into::into)
    }
}

#[async_trait]
impl Seedable for Animal {
    async fn seed_data(db: &sqlx::PgPool, animals: &[Self]) -> ModelResult<()> {
        for animal in animals {
            sqlx::query("INSERT INTO animals
                (id, pid, organisation_pid, tag_id, name, specie_id, breed_id, date_of_birth,
                gender, parent_female_id, parent_male_id, status, purchase_date, purchase_price,
                weight_at_birth, current_weight, notes, created_by, created_at, updated_at)
                VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20)")
                .bind(animal.id)
                .bind(animal.pid)
                .bind(animal.organisation_pid)
                .bind(animal.tag_id.as_str())
                .bind(animal.name.as_str())
                .bind(animal.specie_id)
                .bind(animal.breed_id)
                .bind(animal.date_of_birth)
                .bind(animal.gender.as_str())
                .bind(animal.parent_female_id)
                .bind(animal.parent_male_id)
                .bind(animal.status.as_str())
                .bind(animal.purchase_date)
                .bind(animal.purchase_price)
                .bind(animal.weight_at_birth)
                .bind(animal.current_weight)
                .bind(animal.notes.as_deref())
                .bind(animal.created_by)
                .bind(animal.created_at)
                .bind(animal.updated_at)
                .execute(db)
                .await?;
        }
        Ok(())
    }
}
