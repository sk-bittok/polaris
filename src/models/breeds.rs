#![allow(clippy::missing_errors_doc)]

use std::borrow::Cow;

use chrono::{DateTime, FixedOffset};
use serde::{Deserialize, Serialize};
use sqlx::{Encode, Executor, Postgres, prelude::FromRow};
use uuid::Uuid;

use super::{ModelError, ModelResult};

#[derive(Debug, Clone, Deserialize)]
pub struct BreedQuery {
    specie: Option<String>,
    name: Option<String>,
}

#[derive(Debug, Deserialize, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RegisterBreed<'a> {
    pub(crate) specie: Cow<'a, str>,
    pub(crate) name: Cow<'a, str>,
    pub(crate) description: Option<Cow<'a, str>>,
    pub(crate) typical_male_weight_range: Option<Cow<'a, str>>,
    pub(crate) typical_female_weight_range: Option<Cow<'a, str>>,
    pub(crate) typical_gestation_period: Option<Cow<'a, str>>,
}

#[derive(Debug, Deserialize, Serialize, Clone, FromRow, Encode)]
pub struct Breed {
    pub(crate) id: i32,
    pub(crate) specie: String,
    pub(crate) name: String,
    pub(crate) description: Option<String>,
    pub(crate) typical_male_weight_range: Option<String>,
    pub(crate) typical_female_weight_range: Option<String>,
    pub(crate) typical_gestation_period: Option<String>,
    pub(crate) created_at: DateTime<FixedOffset>,
    pub(crate) organisation_pid: Option<Uuid>,
    pub(crate) is_system_defined: bool,
}

impl Breed {
    pub async fn find_by_id<'e, C>(db: C, id: i32, org_pid: Uuid) -> ModelResult<Self>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query_as::<_, Self>("SELECT * FROM breeds WHERE id = $1 AND (is_system_defined = TRUE OR organisation_pid = $2)")
            .bind(id)
            .bind(org_pid)
            .fetch_optional(db)
            .await?
            .ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn find_by_all<'e, C>(
        db: C,
        org_pid: Uuid,
        conditions: &BreedQuery,
    ) -> ModelResult<Vec<Self>>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let mut query =
            sqlx::query_as::<_, Self>("SELECT * FROM breeds WHERE  (is_system_defined = TRUE OR organisation_pid = $1) ORDER BY name")
                .bind(org_pid);

        if let Some(name) = &conditions.name {
            query = sqlx::query_as::<_, Self>(
                "SELECT * FROM breeds WHERE (is_system_defined = TRUE OR organisation_pid = $1) AND name LIKE $2",
            )
            .bind(org_pid)
            .bind(format!("%{name}%"));
        }

        if let Some(specie) = &conditions.specie {
            query = sqlx::query_as::<_, Self>(
                "SELECT * FROM breeds WHERE organisation_pid = $1 AND specie LIKE $2",
            )
            .bind(org_pid)
            .bind(format!("%{specie}%",));
        }

        query.fetch_all(db).await.map_err(Into::into)
    }

    pub async fn create<'e, C>(db: C, org_pid: Uuid, dto: &RegisterBreed<'_>) -> ModelResult<Self>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let query = sqlx::query_as::<_, Self>(
            "INSERT INTO breeds (specie, name, typical_male_weight_range, typical_female_weight_range, typical_gestation_period, organisation_pid, description)
                VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING *
            ")
        .bind(&dto.specie)
        .bind(&dto.name)
        .bind(&dto.typical_male_weight_range)
        .bind(&dto.typical_female_weight_range)
        .bind(&dto.typical_gestation_period)
        .bind(org_pid)
        .bind(&dto.description)
        .fetch_one(db)
        .await?;

        Ok(query)
    }
}
