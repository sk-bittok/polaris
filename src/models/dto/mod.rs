pub mod auth;

pub use self::auth::*;
