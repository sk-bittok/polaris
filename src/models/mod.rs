pub mod animals;
pub mod breeds;
pub mod dto;
pub mod enums;
pub mod errors;
pub mod health;
pub mod livestock;
pub mod orgs;
pub mod production;
pub mod roles;
pub mod species;
pub mod summaries;
pub mod users;
pub mod weight;

pub use self::{
    errors::{ModelError, ModelResult},
    summaries::*,
};
