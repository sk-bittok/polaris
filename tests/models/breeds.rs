#![allow(unused)]
use insta::{Settings, assert_debug_snapshot, with_settings};
use polaris::models::{
    breeds::{Breed, BreedQuery},
    dto::{RegisterBreed, UpdateBreed},
    species::Specie,
};
use rstest::rstest;
use serial_test::serial;
use uuid::Uuid;

use crate::{boot_test, cleanup_date, cleanup_int, cleanup_uuid, seed_data};

macro_rules! configure_insta {
    ($(expr:expr),*) => {
        let mut settings = Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("breeds");
        settings.set_snapshot_path("snapshots/breeds");
        let _guard = settings.bind_to_scope();
    };
}

#[rstest]
#[case(
    "can_find_all_no_query",
    BreedQuery { specie: None, name: None }
)]
#[case(
    "can_find_all_specie_query",
    BreedQuery { specie: Some("cattle".into()), name: None }
)]
#[case(
    "can_find_all_name_query",
    BreedQuery { specie: None, name: Some("Hereford".into()) }
)]
#[tokio::test]
#[serial]
async fn can_find_all(#[case] test_name: &str, #[case] query: BreedQuery) {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let org_pid = Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap();

    let result = Breed::find_by_all(&ctx.db, org_pid, &query).await;

    with_settings!(
        {
            filters => {
                let mut filters = cleanup_date().to_vec();
                filters.extend(cleanup_uuid().to_vec());
                filters.extend(cleanup_int().to_vec());
                filters
            }
        },
        {
            assert_debug_snapshot!(test_name, result);
        }
    );
}

#[tokio::test]
#[serial]
async fn can_find_by_id() {
    configure_insta!();
    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let org_pid = Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap();

    let result = Breed::find_by_id(&ctx.db, 1, org_pid).await;

    with_settings!(
        {
            filters => {
                let mut filters = cleanup_date().to_vec();
                filters.extend(cleanup_uuid().to_vec());
                filters.extend(cleanup_int().to_vec());
                filters
            }
        },
        {
            assert_debug_snapshot!( result);
        }
    );
}

#[tokio::test]
#[serial]
async fn can_create_one() {
    configure_insta!();
    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let org_pid = Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap();

    let mut params = RegisterBreed::new("cattle", "Sussex");

    params = params.male_weight_range("950-1050");
    params = params.female_weight_range("550-670");
    params = params.gestation_period("280 days");
    params = params.description("A dark red cattle breed kept for meat");

    let specie = Specie::find_by_name(&ctx.db, params.specie())
        .await
        .unwrap();

    let result = Breed::create(&ctx.db, org_pid, &params, specie.id()).await;

    with_settings!(
        {
            filters => {
                let mut filters = cleanup_date().to_vec();
                filters.extend(cleanup_uuid().to_vec());
                filters.extend(cleanup_int().to_vec());
                filters
            }
        },
        {
            assert_debug_snapshot!( result);
        }
    );
}

#[tokio::test]
#[serial]
async fn can_delete_by_id() {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let org_pid = Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap();

    let result = Breed::delete_breed(&ctx.db, org_pid, 100).await;

    assert_debug_snapshot!(result);
}

#[tokio::test]
#[serial]
async fn can_update_by_id() {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let org_pid = Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap();

    let id = 100;

    let mut params = UpdateBreed::default();

    params = params.description("It is typically light brown in colour, though this can range from being almost grey to dull black, which is known as Mulberry. 
        They can also have white patches which may cover much of the animal. A true Jersey will however always have a black nose bordered by an almost white muzzle.
    The Jersey hard black feet are much less prone to lameness. The Jersey is relatively small in size - about 400 to 450kgs in weight and have a fine but strong frame.");

    params = params.gestation_period("269 days");
    params = params.male_weight_range("550-700 kg");
    params = params.female_weight_range("400-450 kg");

    let result = Breed::update_by_id(&ctx.db, org_pid, id, &params).await;

    with_settings!({
        filters => {
             crate::cleanup_date().to_vec()
        }
    }, {
        assert_debug_snapshot!(result);
    })
}
