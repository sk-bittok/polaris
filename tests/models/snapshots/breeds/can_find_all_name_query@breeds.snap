---
source: tests/models/breeds.rs
assertion_line: 58
expression: result
---
Ok(
    [
        Breed {
            id: ID,
            specie_id: ID,
            name: "Hereford",
            description: Some(
                "Beef cattle spotting a red coat with a white head and socks",
            ),
            typical_male_weight_range: Some(
                "1000-1045",
            ),
            typical_female_weight_range: Some(
                "545-680",
            ),
            typical_gestation_period: Some(
                "285 days",
            ),
            created_at: DATE,
            organisation_pid: None,
            is_system_defined: true,
        },
    ],
)
